### PR TITLE
A heuristic to flush only a fraction of the queue when we are out of space for a new kernel

### PR DESCRIPTION
**Pull request by sryap**
When the capacity of the asyncOps vector reaches the
MAX_INFLIGHT_COMMANDS_PER_QUEUE threshold, flush only the first
fraction of the vector to overlap the cleanup process (operator
nullification and GC in removeAsyncOp()) with the ongoing GPU
computation.  The fraction size is tunable via the environment
variable HCC_QUEUE_FLUSHING_RATIO.  The default fraction size is half
of the queue.

Signed-off-by: Sarunya Pumma <sarunya.pumma@amd.com>

### CODE REVIEW

**Comment by whchung** [comment]
Empirical benchmark results shows ~6% wall clock time improvement for resnet50 model on TensorFlow 1.3 on 1GPU. ~12% improvement on 4GPUs.

**Comment by whchung** [comment]
@sarunyap will need your help to provide more background information wrt to this PR, and how it could be improved further.

File: lib/hsa/mcwamp_hsa.cpp [hunk 1]
```diff
@@ -1556,8 +1555,13 @@ class HSAQueue final : public KalmarQueue
 
 
         bool foundFirstValidOp = false;
+        int oldAysncOpsSize = asyncOps.size();
+        int numWaitOps = oldAysncOpsSize - 1;
+        if (drainingQueue_) {
+            numWaitOps = oldAysncOpsSize / 2;
```
Comment (whchung) [inline lib/hsa/mcwamp_hsa.cpp]: Please notice this is just a heuristic and might have to be tuned per application. So maybe it's better to expose it as an environment variable.

Comment (sryap) [reply lib/hsa/mcwamp_hsa.cpp]: @whchung I have added the environment variable, namely HCC_QUEUE_FLUSHING_RATIO, to tune the percentage of the queue to be flushed when the queue is full.  The percentage has to be greater than zero.  Any invalid value will be set to the default value.  The default HCC_QUEUE_FLUSHING_RATIO is 50, which is 50 percent of the queue.

**Review by scchan — CHANGES_REQUESTED** [review]
Wouldn't it be simpler to adjust the MAX_INFLIGHT_COMMANDS_PER_QUEUE limit rather than introducing new logic for this?

**Comment by whchung** [comment]
@sarunyap can chime in to share more details but to my understanding enlarging max infight commands in the queue may actually cause more siginificant perf loss because draining the queue now takes longer time and GPU would be idling. Perhaps we need empirical data from @sarunyap

**Comment by scchan** [comment]
I mean you could lower the limit so that the drainqueue flag gets set sooner and there will be less ops to drain?

**Comment by whchung** [comment]
@scchan thanks! That’s indeed a valid argument. I think we do need experiment data then @sarunyap

**Comment by sryap** [comment]
@scchan @whchung Thank you for reviewing my changes.  I think lowering the limit of MAX_INFLIGHT_COMMANDS_PER_QUEUE will not solve the problem because HCC will still wait for all operators in the queue to complete and the cleanup (removeAsyncOp()) of the last operator in the queue will be done when GPU is sitting idle.  This will only make the cleanup faster, but the cleanup time will still show up as a gap in the total time.

What my changes do is to hide the cleanup process of the last operator that we wait behind the ongoing GPU computation.

**Comment by bensander** [comment]
Right we want to avoid draining the queue to zero so we can reclaim some of the resources but also keep the gpu doing useful work

**Review by scchan — APPROVED** [review]

**Comment by scchan** [comment]
@whchung any concern or any additional discussion needed for this to be merged?

**Comment by whchung** [comment]
@scchan sorry for belated reply. no concern here.

**Pull request merged**